src/engraving/dom/fret.cpp

```diff
@@ -529,12 +566,65 @@ void FretDiagram::removeDotsMarkers(int ss, int es, int fret)
     }
 }
 
+void FretDiagram::applyDiagramPattern(FretDiagram* diagram, const String& pattern)
+{
+    diagram->clear();
+
+    int strings = static_cast<int>(pattern.size());
+
+    diagram->setStrings(strings);
+    diagram->setFrets(4);
+    diagram->setPropertyFlags(Pid::FRET_STRINGS, PropertyFlags::UNSTYLED);
+    diagram->setPropertyFlags(Pid::FRET_FRETS,   PropertyFlags::UNSTYLED);
+    int offset = 0;
+    int barreString = -1;
+    std::vector<std::pair<int, int> > dotsToAdd;
+
+    for (int i = 0; i < strings; i++) {
+        Char c = pattern.at(i);
+        if (c == 'X' || c == 'O') {
+            FretMarkerType mt = (c == 'X' ? FretMarkerType::CROSS : FretMarkerType::CIRCLE);
+            diagram->setMarker(i, mt);
+        } else if (c == '-' && barreString == -1) {
+            barreString = i;
+        } else {
+            int fret = c.digitValue();
+            if (fret != -1) {
+                dotsToAdd.push_back(std::make_pair(i, fret));
+                if (fret - 3 > 0 && offset < fret - 3) {
+                    offset = fret - 3;
+                }
+            }
+        }
+    }
+
+    if (offset > 0) {
+        diagram->setFretOffset(offset);
+    }
+
+    for (std::pair<int, int> d : dotsToAdd) {
+        diagram->setDot(d.first, d.second - offset, true);
+    }
+
+    // This assumes that any barre goes to the end of the fret
+    if (barreString >= 0) {
+        diagram->setBarre(barreString, -1, 1);
+    }
+}
+
 //---------------------------------------------------------
 //   clear
 //---------------------------------------------------------
 
 void FretDiagram::clear()
 {
+    m_strings = 6;
```

Comment:
I don't really like that now we have these values in 2 different places. What we should do instead:

1. initialize these members (i.e. in fret.h, when declaring them) with some "undefined" values, like 0 and false
2. add a function like initDefaultValues (or rename clear) and move this code into it